packages/smithy-http/src/smithy_http/interceptors/user_agent.py

```diff
@@ -0,0 +1,151 @@
+    ) -> HTTPRequest:
+        user_agent = context.properties["user_agent"]
+        request = context.transport_request
+        request.fields.set_field(Field(name="User-Agent", values=[str(user_agent)]))
+        return context.transport_request
+
+
+_USERAGENT_ALLOWED_OS_NAMES = (
+    "windows",
+    "linux",
+    "macos",
+    "android",
+    "ios",
+    "watchos",
+    "tvos",
+    "other",
+)
+_USERAGENT_PLATFORM_NAME_MAPPINGS = {"darwin": "macos"}
+_USERAGENT_SDK_NAME = "python"
+
+
+class UserAgentBuilder:
+    def __init__(
+        self,
+        *,
+        platform_name: str | None = None,
+        platform_version: str | None = None,
+        platform_machine: str | None = None,
+        python_version: str | None = None,
+        python_implementation: str | None = None,
+        sdk_version: str | None = None,
+    ) -> None:
+        self._platform_name = platform_name
+        self._platform_version = platform_version
+        self._platform_machine = platform_machine
+        self._python_version = python_version
+        self._python_implementation = python_implementation
+        self._sdk_version = sdk_version
+
+    @classmethod
+    def from_environment(cls) -> Self:
+        return cls(
+            platform_name=platform.system(),
+            platform_version=platform.release(),
+            platform_machine=platform.machine(),
+            python_version=platform.python_version(),
+            python_implementation=platform.python_implementation(),
+            sdk_version=smithy_core.__version__,
+        )
+
+    def build(self) -> UserAgent:
+        user_agent = UserAgent()
+        user_agent.sdk_metadata.extend(self._build_sdk_metadata())
+        user_agent.ua_metadata.append(UserAgentComponent(prefix="ua", name="2.1"))
+        user_agent.os_metadata.extend(self._build_os_metadata())
+        user_agent.os_metadata.extend(self._build_architecture_metadata())
+        user_agent.language_metadata.extend(self._build_language_metadata())
+
+        return user_agent
+
+    def _build_sdk_metadata(self) -> list[UserAgentComponent]:
+        if self._sdk_version:
+            return [
+                UserAgentComponent(prefix=_USERAGENT_SDK_NAME, name=self._sdk_version)
+            ]
+        return []
+
+    def _build_os_metadata(self) -> list[UserAgentComponent]:
+        """Build the OS/platform components of the User-Agent header string.
+
+        For recognized platform names that match or map to an entry in the list
+        of standardized OS names, a single component with prefix "os" is
+        returned. Otherwise, one component "os/other" is returned and a second
+        with prefix "md" and the raw platform name.
+
+        String representations of example return values:
+         * ``os/macos#10.13.6``
+         * ``os/linux``
+         * ``os/other``
+         * ``os/other md/foobar#1.2.3``
+        """
+        if self._platform_name is None or self._platform_name == "":
```

Comment:
nit; we can consolidate this into:
```suggestion
        if self._platform_name in (None, ""):
```